prov/util/src/util_ep.c

```diff
@@ -36,6 +36,44 @@
 #include <fi_enosys.h>
 #include <fi_util.h>
 
+int ofi_ep_bind_cq(struct util_ep *ep, struct util_cq *cq, uint64_t flags)
+{
+	const struct fi_provider *prov = ep->domain->fabric->prov;
+
+	if (flags & ~(FI_TRANSMIT | FI_RECV |
+		      FI_SELECTIVE_COMPLETION))
+		FI_WARN(prov, FI_LOG_EP_CTRL,
+			"Unsupported flags\n");
+		return -FI_EBADFLAGS;
+
+	if (((flags & FI_TRANSMIT) && ep->tx_cq) ||
+	    ((flags & FI_RECV) && ep->rx_cq))
+		FI_WARN(prov, FI_LOG_EP_CTRL,
+			"Duplicate CQ binding\n");
+		return -FI_EINVAL;
+
+	if (flags & FI_TRANSMIT) {
+		ep->tx_cq = cq;
+		if (!(flags & FI_SELECTIVE_COMPLETION))
+			ep->tx_op_flags |= FI_COMPLETION;
+		ofi_atomic_inc32(&cq->ref);
+	}
+
+	if (flags & FI_RECV) {
+		ep->rx_cq = cq;
+		if (!(flags & FI_SELECTIVE_COMPLETION))
+			ep->rx_op_flags |= FI_COMPLETION;
+		ofi_atomic_inc32(&cq->ref);
+	}
+
+	if (flags & (FI_TRANSMIT | FI_RECV)) {
```

Comment:
This if statement changes the behavior for the udp provider, which only adds the ep to the receive cq list.  That could be the reason why that provider is failing in travis.  (udp may only be cleaning up the receive cq.)  It results in slightly more overhead for udp, since the underlying socket will be processed for receives twice, once coming from the send cq.  For udp, please revert those changes and have it keep its own bind call.

To keep some code sharing, maybe separate out the bind validation check, so that this call and the udp call can both use them?